### PR TITLE
Added Bitstamp support

### LINKED ISSUE #436: Bitstamp support
## Abstract

Bitstamp support for trades import. As one of the oldest exchanges it probably has many users.

## Motivation

/

## Specification

/

### PR DESCRIPTION
**Pull request by vnavascues**
Pending:
- [x] Wait for Bitstamp support answers regarding the deposit/withdrawal JSON, and nail down `_deserialize_asset_movement()`.
- [x] Test suite for asset movements.
- [x] Wait for Bitstamp support answers regarding requests time limit.

Related Bitstamp API clients:
https://godoc.org/github.com/thrasher-/gocryptotrader/exchanges/bitstamp#UserTransactions
https://github.com/kmadac/bitstamp-python-client/blob/master/bitstamp/client.py#L256

Closes #436

### CODE REVIEW

File: rotkehlchen/data/importer.py [hunk 1]
```diff
@@ -55,6 +55,9 @@ def exchange_row_to_location(entry: str) -> Location:
     # TODO: Check if this is the correct string for Gemini from cointracking
     elif entry == 'Gemini':
         return Location.GEMINI
+    # ! TODO VN: add Bitstamp support
```
Comment (vnavascues) [inline rotkehlchen/data/importer.py:58]: @LefterisJP I'd like to know if this is right.

File: rotkehlchen/exchanges/bitstamp.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
```
Comment (vnavascues) [inline rotkehlchen/exchanges/bitstamp.py:645]: @LefterisJP I've implemented this method for reducing the blocks related with processing unsuccessful requests, and standardize it a bit more (sometimes between exchanges differ in some logs up or down). Also cause it has really helped me to see the subtle differences between methods, and it has been much easier to debug and test.

File: rotkehlchen/exchanges/bitstamp.py [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+                    result_timestamp = iso8601ts_to_timestamp(raw_result['datetime'])
+
+                    if result_timestamp > end_ts:
+                        is_result_timesamp_gt_end_ts = True  # prevent extra request
+                        break
+
+                    if case == 'trades':
+                        result = self._deserialize_trade(raw_result)
+                    elif case == 'asset_movements':
+                        # TODO: fix this type...
```
Comment (vnavascues) [inline rotkehlchen/exchanges/bitstamp.py:420]: I have a type error here. I've tried to declare `result: Union[Trade, AssetMovement]` above but nothing.

File: rotkehlchen/exchanges/bitstamp.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+                self.msg_aggregator.add_error(
+                    f'Got remote error while querying Bistamp trades: {msg}',
+                )
+                no_results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore
+                return no_results
+
+            has_results = False
+            is_result_timesamp_gt_end_ts = False
+            for raw_result in response_list:
+                if (
```
Comment (vnavascues) [inline rotkehlchen/exchanges/bitstamp.py:404]: Skipping by `type` only applies to the "user_transactions" endpoint.
We should be able to use this function for paginating other endpoints at least via limit/offset/datetime, but also any other particular endpoint attribute.

File: rotkehlchen/exchanges/bitstamp.py [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+        """Process a deposit/withdrawal user transaction from Bitstamp and
+        deserialize it.
+
+        Can raise Can raise DeserializationError.
+
+        From Bitstamp documentation, deposits/withdrawals can have a fee
+        (the amount is expected to be in the currency involved)
+        https://www.bitstamp.net/fee-schedule/
+
+        TODO: WIP, waiting for Binance Support clarification about the JSON
```
Comment (vnavascues) [inline rotkehlchen/exchanges/bitstamp.py:509]: As I've explained we will need some clarification about the JSON.

**Comment by vnavascues** [comment]
I will have to look at this, some logs I have running it.

```shell
[03/12/2020 18:51:06 GMT] ERROR rotkehlchen.api.rest: Greenlet for task 11 dies with exception: near "AND": syntax error.
Exception Name: <class 'pysqlcipher3.dbapi2.OperationalError'>
Exception Info: near "AND": syntax error
Traceback:
   File "src/gevent/greenlet.py", line 818, in gevent._greenlet.Greenlet.run
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/api/rest.py", line 236, in _do_query_async
    result = getattr(self, command)(**kwargs)
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/api/rest.py", line 541, in _get_trades
    trades = self.rotkehlchen.query_trades(from_ts=from_ts, to_ts=to_ts, location=location)
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/rotkehlchen.py", line 597, in query_trades
    trades = self.query_location_trades(from_ts, to_ts, location)
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/rotkehlchen.py", line 663, in query_location_trades
    location_trades = exchange.query_trade_history(start_ts=from_ts, end_ts=to_ts)
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/utils/interfaces.py", line 115, in wrapper
    result = f(wrappingobj, *args, **kwargs)
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/exchanges/exchange.py", line 155, in query_trade_history
    end_ts=query_end_ts,
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/exchanges/bitstamp.py", line 225, in query_online_trade_history
    location=Location.BITSTAMP,
  File "/home/victor/Projects/rotki_vnavascues/rotkehlchen/db/dbhandler.py", line 2183, in get_trades
    results = cursor.execute(query, bindings)
```

**Comment by vnavascues** [comment]
@LefterisJP  Re. Bitstamp support
```
Please note that our request limit also applies to v2 API endpoints and is, as you mentioned, 8000 requests per 10 minutes.
```

File: rotkehlchen/exchanges/bitstamp.py [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+}
+# Max limit for all API v2 endpoints
+API_MAX_LIMIT = 1000
+# user_transactions endpoint constants
+# Sort mode
+USER_TRANSACTION_SORTING_MODE = 'asc'
+# Starting `since_id`
+USER_TRANSACTION_MIN_SINCE_ID = 1
+# Trade type int
+USER_TRANSACTION_TRADE_TYPE = {14}
```
Comment (vnavascues) [inline rotkehlchen/exchanges/bitstamp.py:74]: @LefterisJP this is wrong, it should be `2`, my bad.

File: rotkehlchen/exchanges/bitstamp.py [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+                continue
+
+            asset_balance[asset] = Balance(
+                amount=amount,
+                usd_value=amount * usd_price,
+            )
+
+        return asset_balance, ''
+
+    def query_online_deposits_withdrawals(
```
Comment (vnavascues) [inline rotkehlchen/exchanges/bitstamp.py:168]: @LefterisJP at the moment it only requests the `user_transactions` endpoint, which is not capable of returning the asset movement `address`, nor `transaction_id`. These would require to request as well `crypto_transactions` and do a mix. At the moment I set both AssetMovement attributes to None, as other exchanges do, and I'm not sure if we really want to request both endpoints and mix just for getting these 2 attributes.

You can search "crypto transactions" (below "user transactions") in their docs. Below an image.

image

File: rotkehlchen/exchanges/bitstamp.py [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+        return None
+
+    def _deserialize_asset_movement(
+            self,
+            raw_movement: Dict[str, Any],
+    ) -> AssetMovement:
+        """Process a deposit/withdrawal user transaction from Bitstamp and
+        deserialize it.
+
+        Can raise Can raise DeserializationError.
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:475]: Remove extra "Can raise"

File: rotkehlchen/exchanges/bitstamp.py (hunk 5 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/bitstamp.py:481]: It seems you got it now? Can you verify that your logic is correct from the response?

File: rotkehlchen/exchanges/bitstamp.py [hunk 9]
```diff
@@ -0,0 +1,711 @@
...
+            raise DeserializationError(
+                'Could not deserialize Bitstamp asset movement from user transaction. '
+                f'Unexpected asset amount combination found in: {raw_movement}.',
+            )
+
+        asset_movement = AssetMovement(
+            timestamp=timestamp,
+            location=Location.BITSTAMP,
+            category=category,
+            address=None,  # requires query "crypto_transactions" endpoint
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:523]: So you need to do an extra query to get the address and the transaction hash?

File: rotkehlchen/exchanges/bitstamp.py [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+        )
+        return asset_movement
+
+    def _deserialize_trade(
+            self,
+            raw_trade: Dict[str, Any],
+    ) -> Trade:
+        """Process a trade user transaction from Bitstamp and deserialize it.
+
+        Can raise Can raise DeserializationError.
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:539]: Remove extra "Can raise"

File: rotkehlchen/exchanges/bitstamp.py [hunk 11]
```diff
@@ -0,0 +1,711 @@
...
+                f'Trade pair not found in: {raw_result}.',
+            )
+
+        # NB: `pair_get_assets()` is not used for simplifying the calls and
+        # storing the raw pair strings.
+        base_asset_symbol, quote_asset_symbol = pair.split('_')
+        try:
+            base_asset = Asset(base_asset_symbol)
+            quote_asset = Asset(quote_asset_symbol)
+        except UnknownAsset as e:
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:598]: What about `UnsupportedAsset `exception? Shouldn't it also be caught here?

File: rotkehlchen/exchanges/bitstamp.py [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+            self,
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: str,
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:648]: The type here should be: `Literal[combination of all literals above]`

File: rotkehlchen/exchanges/bitstamp.py (hunk 2 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/bitstamp.py:645]: Sure no problem. It looks nice and feels neat to have this logic abstracted away in a function like that.

File: rotkehlchen/exchanges/bitstamp.py [hunk 13]
```diff
@@ -0,0 +1,711 @@
...
+            elif case in {'trades', 'asset_movements'}:
+                self.msg_aggregator.add_error(
+                    f'Got remote error while querying Bistamp {case_pretty}: {msg}',
+                )
+                no_results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore
+                return no_results
+            else:
+                raise AssertionError(f'Unexpected Bitstamp response_case: {case}.')
+
+        # Get error code
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:678]: A request:

1. Please don't add unnecessary whitespace. There is no need for whitespace before and after a function call.
2. Don't add unneded comments:
    - Get error code is not needed. It's obvious you are doing it.
    - Local system clock not synch not needed. The function is named perfectly well.

File: rotkehlchen/exchanges/bitstamp.py [hunk 14]
```diff
@@ -0,0 +1,711 @@
...
+        error_code = response_dict.get('code', None)
+
+        # Local system clock not synch
+        self._check_for_system_clock_not_synced_error(error_code)
+
+        # Errors related with the API key return a human redable message
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:684]: redable -> readable

File: rotkehlchen/exchanges/bitstamp.py [hunk 15]
```diff
@@ -0,0 +1,711 @@
...
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: str,
+    ) -> Union[
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:649]: Just a note (don't change anything): Not a big fan of the amount of different possible returns of this function.

File: rotkehlchen/exchanges/bitstamp.py [hunk 16]
```diff
@@ -0,0 +1,711 @@
...
+        # Custom response per case
+        if case == 'validate_api_key':
+            raise RemoteError(msg)
+        if case == 'balances':
+            return None, msg
+        elif case in {'trades', 'asset_movements'}:
+            self.msg_aggregator.add_error(
+                f'Got remote error while querying Bistamp {case_pretty}: {msg}',
+            )
+            no_results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:708]: Both here and above where you do the same thing there is no need tospecify this type. You don't ever return any trades or asset movements from here. The error is correct. Try this:

```diff
diff --git a/rotkehlchen/exchanges/bitstamp.py b/rotkehlchen/exchanges/bitstamp.py
index f055312a..d3d16349 100644
--- a/rotkehlchen/exchanges/bitstamp.py
+++ b/rotkehlchen/exchanges/bitstamp.py
@@ -647,8 +647,7 @@ class Bitstamp(ExchangeInterface):
             response: Response,
             case: str,
     ) -> Union[
-        List[Trade],
-        List[AssetMovement],
+        List,
         Tuple[bool, str],
         Tuple[Optional[Dict[Asset, Balance]], str],
     ]:
@@ -670,8 +669,7 @@ class Bitstamp(ExchangeInterface):
                 self.msg_aggregator.add_error(
                     f'Got remote error while querying Bistamp {case_pretty}: {msg}',
                 )
-                no_results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore
-                return no_results
+                return []
             else:
                 raise AssertionError(f'Unexpected Bitstamp response_case: {case}.')
 
@@ -705,7 +703,6 @@ class Bitstamp(ExchangeInterface):
             self.msg_aggregator.add_error(
                 f'Got remote error while querying Bistamp {case_pretty}: {msg}',
             )
-            no_results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore
-            return no_results
+            return []
         else:
             raise AssertionError(f'Unexpected Bitstamp response_case: {case}.')
```

File: rotkehlchen/tests/exchanges/test_bitstamp.py [hunk 17]
```diff
@@ -0,0 +1,724 @@
...
+            Asset('LINK'): Balance(
+                amount=FVal('1000'),
+                usd_value=FVal('1500'),
+            ),
+        }
+        assert msg == ''
+
+
+# method: `_deserialize_trade()`
+def test_deserialize_trade__buy(mock_bitstamp):
```
Comment (LefterisJP) [inline rotkehlchen/tests/exchanges/test_bitstamp.py:241]: here and below why double underscore after `trade`?

File: rotkehlchen/tests/exchanges/test_bitstamp.py [hunk 18]
```diff
@@ -0,0 +1,724 @@
...
+        fee=FVal("0.00610000"),
+        fee_currency=Asset('EUR'),
+        link='5',
+        notes='',
+    )
+    trade = mock_bitstamp._deserialize_trade(raw_trade)
+    assert trade == expected_trade
+
+
+# method: `_api_query_paginated()` for 'user_transactions' endpoint and trades
```
Comment (LefterisJP) [inline rotkehlchen/tests/exchanges/test_bitstamp.py:295]: Why are you adding these comments above the tests? It should be obvious from the test name which function is being tested. Don't add unnecessary comments

**Review by LefterisJP — COMMENTED** [review]
Hey @vnavascues so this is my first review of the PR.

Looking good, code looks mostly nice and clean. Thank you.

Nice amount of tests also written (perhaps even too much -- as you come to understand we need to be moving faster, so perhaps if writing so many tests makes you go slow then reduce it a bit)

I left a few comments but nothing major. What I would like to know is if you made the changes after the bitstamp support response for asset movements.

And then if all else is addressed, you can also add a changelog entry and ping @kelsos to add the very little amount of work needed in the frontend.

File: rotkehlchen/exchanges/bitstamp.py (hunk 7 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/bitstamp.py:168]: > and I'm not sure if we really want to request both endpoints and mix just for getting these 2 attributes.

Please make an issue about us doingthis at some point. Does not need to be part of this PR, but we can have an issue to remind us to improve the implementation.

File: rotkehlchen/exchanges/bitstamp.py (hunk 16 above)
Comment (vnavascues) [reply rotkehlchen/exchanges/bitstamp.py]: good one, I tried many things not a single one the right one.

File: rotkehlchen/tests/exchanges/test_bitstamp.py (hunk 17 above)
Comment (vnavascues) [reply rotkehlchen/tests/exchanges/test_bitstamp.py]: I tend to use underscores when two of more tests share the first part of the title (testing the same function) but apply to different logic. It helps separating words/terms that can already contain a `_`. I can stop it :P

File: rotkehlchen/exchanges/bitstamp.py (hunk 4 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/bitstamp.py]: I see you already removed it. Good.

File: rotkehlchen/exchanges/bitstamp.py (hunk 7 above)
Comment (LefterisJP) [reply rotkehlchen/exchanges/bitstamp.py:181]: Issue is here: https://github.com/rotki/rotki/issues/1863

Resolving comment.

File: rotkehlchen/tests/exchanges/test_bitstamp.py (hunk 17 above)
Comment (LefterisJP) [reply rotkehlchen/tests/exchanges/test_bitstamp.py]: It's not bad, but since we don't do it in the rest of the code it's better if you also don't do it. After all a consistent looking codebase is important.

File: rotkehlchen/exchanges/bitstamp.py [hunk 19]
```diff
@@ -0,0 +1,715 @@
...
+
+    def _deserialize_trade(
+            self,
+            raw_trade: Dict[str, Any],
+    ) -> Trade:
+        """Process a trade user transaction from Bitstamp and deserialize it.
+
+        Can raise DeserializationError.
+        """
+        timestamp = iso8601ts_to_timestamp(raw_trade['datetime'])
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/bitstamp.py:554]: This is actually a mistake. I realized it only thanks to the test you yourself wrote @vnavascues. (It fails on my system since I am not in UTC).

`iso8601ts_to_timestamp()` interprets the datetime string to be in local time. We had similar problem with poloniex. So I am pushing a fix for this before merging.

**Review by LefterisJP — COMMENTED** [review]
LGTM. Would love some people with more trades in the exchange to test it and tell us.

**Review by LefterisJP — APPROVED** [review]
